anndata/_core/anndata.py

```diff
@@ -1796,10 +1430,13 @@ def copy(self, filename: Optional[PathLike] = None) -> 'AnnData':
                 varm=self.varm.copy(),
                 obsp=self.obsp.copy(),
                 varp=self.varp.copy(),
-                raw=None if self._raw is None else self._raw.copy(),
                 layers=self.layers.copy(),
                 dtype=dtype,
             )
+            if self._raw is not None:
+                adata._raw = self._raw.copy()
+                adata._raw._adata = adata
+            return adata
```

Comment:
This should still go through the AnnData constructor.

Was there a reason you didn't want to do that?

Reply:
Yes, e.g. `test_sparse_to_dense_inplace` will fail, as  `assert isinstance(backed.raw.X, h5py.Dataset)`:

```pytb
anndata/tests/test_io_conversion.py:62 (test_sparse_to_dense_inplace[scipy-csr])
tmp_path = PosixPath('/tmp/pytest-of-phil/pytest-12/test_sparse_to_dense_inplace_s0')
spmtx_format = <class 'scipy.sparse.csr.csr_matrix'>

    def test_sparse_to_dense_inplace(tmp_path, spmtx_format):
        pth = tmp_path / "adata.h5ad"
        orig = gen_adata((50, 50), spmtx_format)
        orig.raw = orig
        orig.write(pth)
        backed = ad.read_h5ad(pth, backed="r+")
        backed.write(as_dense=("X", "raw/X"))
        new = ad.read_h5ad(pth)
    
        assert_equal(orig, new)
        assert_equal(backed, new)
    
        assert isinstance(new.X, np.ndarray)
        assert isinstance(new.raw.X, np.ndarray)
        assert isinstance(orig.X, spmtx_format)
        assert isinstance(orig.raw.X, spmtx_format)
        assert isinstance(backed.X, h5py.Dataset)
>       assert isinstance(backed.raw.X, h5py.Dataset)
E       AssertionError: assert False
E        +  where False = isinstance(<50x50 sparse matrix of type '<class 'numpy.float32'>'\n	with 1010 stored elements in Compressed Sparse Row format>, <class 'h5py._hl.dataset.Dataset'>)
E        +    where <50x50 sparse matrix of type '<class 'numpy.float32'>'\n	with 1010 stored elements in Compressed Sparse Row format> = <anndata._core.raw.Raw object at 0x7efbf1cd64f0>.X
E        +      where <anndata._core.raw.Raw object at 0x7efbf1cd64f0> = AnnData object with n_obs × n_vars = 50 × 50 backed at '/tmp/pytest-of-phil/pytest-12/test_sparse_to_dense_inplace_s0/...    varm: 'array', 'df', 'sparse'\n    layers: 'array', 'sparse'\n    obsp: 'array', 'sparse'\n    varp: 'array', 'sparse'.raw
E        +    and   <class 'h5py._hl.dataset.Dataset'> = h5py.Dataset

anndata/tests/test_io_conversion.py:80: AssertionError
```

Reply:
That makes it sound like there's something very broken going on. When is `copy` even being called?

I think this needs addressing, since `raw.X` should not be in-memory or sparse here.

Reply:
Out of scope for this PR. If someone want to further fix Raw afterwards, we should do it afterwards.

Reply:
That commit looks right to me. These tests:

```python
assert isinstance(backed.X, h5py.Dataset)
assert isinstance(backed.raw.X, h5py.Dataset)
```

definitely shouldn't fail. I don't see how changing `copy` could make them fail unless something else is going wrong.

Reply:
If I do `self.raw = AnnData(**raw)`, it fails in this way.

Reply:
It makes sense to me that calling copy would make this go into memory. What doesn't make sense to me is that copy is even being called here, and somehow it only effects `.raw`?

Also, how would that even effect these lines, since these are all under a `if not self.isbacked`?

Reply:
> If I do self.raw = AnnData(**raw) as you suggested in the init function, it fails in this way.

Hmm. I don't recall recommending that for the `AnnData.__init__` function, but definitely see how it wouldn't work for backed mode. I'm also unsure of what `**raw` would contain if it were coming from `copy`.

---------------------------

I just tried reverting this block to:

```python
return AnnData(
    X=X,
    obs=self.obs.copy(),
    var=self.var.copy(),
    # deepcopy on DictView does not work and is unnecessary
    # as uns was copied already before
    uns=self.uns.copy()
    if isinstance(self.uns, DictView)
    else deepcopy(self.uns),
    obsm=self.obsm.copy(),
    varm=self.varm.copy(),
    obsp=self.obsp.copy(),
    varp=self.varp.copy(),
    raw=None if self._raw is None else self._raw.copy(),
    layers=self.layers.copy(),
    dtype=dtype,
)
```

on the current commit, and do not get an error.

Reply:
“This block” being `AnnData.copy`?

Reply:
Here is a version of copy that works and uses the initializer:

```python
    def copy(self, filename: Optional[PathLike] = None) -> 'AnnData':
        """Full copy, optionally on disk."""
        if not self.isbacked:
            if self.is_view:
                # TODO: How do I unambiguously check if this is a copy?
                # Subsetting this way means we don't have to have a view type
                # defined for the matrix, which is needed for some of the
                # current distributed backend.
                X = _subset(self._adata_ref.X, (self._oidx, self._vidx)).copy()
            else:
                X = self.X.copy()
            # TODO: Figure out what case this is:
            if X is not None:
                dtype = X.dtype
                if X.shape != self.shape:
                    X = X.reshape(self.shape)
            else:
                dtype = "float32"
            return AnnData(
                X=X,
                obs=self.obs.copy(),
                var=self.var.copy(),
                # deepcopy on DictView does not work and is unnecessary
                # as uns was copied already before
                uns=self.uns.copy()
                if isinstance(self.uns, DictView)
                else deepcopy(self.uns),
                obsm=self.obsm.copy(),
                varm=self.varm.copy(),
                obsp=self.obsp.copy(),
                varp=self.varp.copy(),
                raw=None if self._raw is None else self._raw.copy(),
                layers=self.layers.copy(),
                dtype=dtype,
            )
        else:
            from ..readwrite import read_h5ad

            if filename is None:
                raise ValueError(
                    'To copy an AnnData object in backed mode, '
                    'pass a filename: `.copy(filename=\'myfilename.h5ad\')`.'
                )
            mode = self.file._filemode
            self.write(filename)
            return read_h5ad(filename, backed=mode)
```

Reply:
gotcha!